src/transformStateWithClones.js

```diff
@@ -7,7 +7,45 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let addProperties = {};
+  let removeProperties = {};
+  let cleanProperties = {};
+  let addObject = state;
+
+  const array = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      addProperties = { ...addObject, ...action.extraData };
+      array.push(addProperties);
+      addObject = addProperties;
+      continue;
+    }
+
+    if (action.type === 'removeProperties') {
+      removeProperties = { ...addObject };
+
+      for (const i of action.keysToRemove) {
+        delete removeProperties[i];
+      }
+      array.push(removeProperties);
+      addObject = removeProperties;
```

Comment:
Similar to the 'addProperties' case, you should create a new object inside the 'removeProperties' case to avoid unintended mutations.